src/gtfs.py

```diff
@@ -209,3 +252,42 @@ def add_gtfs_headways(events_df: pd.DataFrame, all_trips: pd.DataFrame, all_stop
         results.append(augmented_events)
 
     return pd.concat(results)
+
+
+current_gtfs_archive = None
+write_gtfs_archive_lock = Lock()
+
+
+def update_current_gtfs_archive_if_necessary():
+    global current_gtfs_archive
+    global write_gtfs_archive_lock
+    with write_gtfs_archive_lock:
+        gtfs_service_date = util.service_date(datetime.datetime.now(util.EASTERN_TIME))
+        needs_update = current_gtfs_archive is None or current_gtfs_archive.service_date != gtfs_service_date
+        if needs_update:
+            if current_gtfs_archive is None:
+                logger.info(f"Downloading GTFS archive for {gtfs_service_date}")
+            else:
+                logger.info(f"Updating GTFS archive from {current_gtfs_archive.service_date} to {gtfs_service_date}")
+            current_gtfs_archive = read_gtfs(gtfs_service_date)
+            # TODO(ian): This will become a per-trip concern in a future change
+            # See https://transitmatters.slack.com/archives/GSJ6F35DW/p1705680401311829?thread_ts=1705677890.833879&cid=GSJ6F35DW
+            disk.write_state({})
+
+
+def get_current_gtfs_archive():
+    global current_gtfs_archive
+    if current_gtfs_archive is None:
+        update_current_gtfs_archive_if_necessary()
+    return current_gtfs_archive
+
+
+def update_gtfs_thread():
```

Comment:
im wondering, how useful is this thread will be at the current frequency? `get_current_gtfs_archive` is called every time a new event is processed, which is about 1-2 times a second, so i think in most instances that function will be the one to trigger an archive update. 

`update_gtfs_thread` most likely will manage to trigger an update only once a day at ~3am when the service date increments, but the first bus of the day can also trigger a similar update.

Reply:
`get_current_gtfs_archive()` will only build the archive if `current_gtfs_archive is None` — it'll happily ignore an out-of-date archive. I included the `update_current_gtfs_archive_if_necessary()` call in case the event processing threads try to peek at it before it gets built.

So in practice, `update_gtfs_thread()` is solely responsible for handling the `service_day` turnover. That way `process_events()` won't block on a new GTFS bundle being built during the first morning trips.

Reply:
ah, makes sense!